lib/class-wp-annotation-utils.php

```diff
@@ -0,0 +1,820 @@
+					),
+				);
+				// @codeCoverageIgnoreEnd
+			} else {
+				$query_vars['meta_query'] = array(
+					'relation' => 'AND',
+					$meta_queries,
+				);
+			}
+		}
+
+		return $query_vars;
+	}
+
+	/**
+	 * Filters WP_Comment_Query clauses.
+	 *
+	 * @since [version]
+	 *
+	 * @param  array            $pieces Array of comment query clauses.
+	 * @param  WP_Comment_Query $query  Current WP_Comment_Query instance.
+	 *
+	 * @return array                    Array of comment query clauses.
+	 *
+	 * @see WP_Annotation_Utils::on_rest_comment_query()
+	 * @see WP_Comment_Query::get_comment_ids()
+	 */
+	public static function on_comments_clauses( $pieces, $query ) {
+		global $wpdb;
+
+		/*
+		 * Annotations can only be returned when 'cache_domain=annotations'. This keeps
+		 * annotations out of any normal comment query. If a plugin wants to query annotations,
+		 * it can set 'cache_domain=annotations' also.
+		 */
+		if ( 'annotations' !== $query->query_vars['cache_domain'] ) {
+			$annotation_types = self::$types;
+
+			foreach ( $annotation_types as &$_type ) {
+				$_type = $wpdb->prepare( '%s', $_type );
+			}
+			$pieces['where'] .= $pieces['where'] ? ' AND ' : '';
+			$pieces['where'] .= 'comment_type NOT IN (' . implode( ', ', $annotation_types ) . ')';
+		}
+
+		/*
+		 * Work around a bug in WP_Comment_Query. If 'hierarchical' is given, WP_Comment_Query
+		 * forces 'parent' to '0', even if 'parent__in' is defined. That's a problem because
+		 * the REST API uses 'parent__in'. See: <https://git.io/vNc8j>
+		 */
+		if ( 'annotations' === $query->query_vars['cache_domain'] && $query->query_vars['hierarchical'] ) {
+			// If performing a hierarchical comment query, WP_Comment_Query will set the 'parent' query var
+			// to a value of 0 by default; i.e., if it wasn't defined already. That conflicts with 'parent__in'.
+			if ( $query->query_vars['parent__in'] && ! $query->query_vars['parent'] ) {
+
+				// So 'parent__in' is not empty, and 'parent' is empty (e.g., 0 or otherwise).
+				// Now, if the clause itself contains 'comment_parent IN' as a result of 'parent__in'.
+				if ( preg_match( '/\bcomment_parent\s+IN\b/i', $pieces['where'] ) ) {
+					// Then remove the conflicting 'comment_parent =' SQL that is brought about by 'parent'.
+					// In other words, we want to do away with 'parent' and let 'parent__in' work as expected.
+					$pieces['where'] = preg_replace( '/\s+AND\s+comment_parent\s*\=\s*[0-9]+/', '', $pieces['where'] );
+				}
+			}
+		}
+
+		return $pieces;
+	}
+
+	/**
+	 * Checks an annotation comment type.
+	 *
+	 * @since [version]
+	 *
+	 * @param  string $type Comment type.
+	 *
+	 * @return bool         True if comment is valid.
+	 */
+	public static function is_valid_type( $type ) {
+		/**
+		 * Filters comment types allowed for annotations.
+		 *
+		 * @since [version]
+		 *
+		 * @param array Comment types allowed for annotations.
+		 */
+		$allow_types = apply_filters( 'annotation_allow_types', self::$types );
```

Comment:
I mean that this filter should be applied in the `on_comments_clauses` method. Otherwise, a type added by this filter will be allowed, but will also be shown in default comment overview. Such as the front-end.

Reply:
Right, I see what you meant now. That will get fixed when I move the filter and change the name. Definitely agree.